# COA Lab:
This is a lab for COA.
We are gonna cover following points in this lab:
1. [N bit adder](#addition)
2. [N bit subtractor](#subtraction)
3. [Partial Product Method of Multiplication](#partial-product-method)
4. [Booth's Multiplication Algorithm](#booths-multiplication-method)
5. [Restoring Division](#restoring-division)
6. [Non restoring Division](#non-restoring-division)
7. [Cache Replacement Algorithm](#cache-replacement-algorithm) <br />
    a. [FIFO](#cache-replacement-algorithm-by-fifo) <br />
    b. [LRU](#cache-replacement-algorithm-by-lru) <br />

### Contributers:
1. [Krischal Khanal (077BEI)](https://github.com/krischal111)

## Logic
The following code segment implements some of the logic required for other upcoming codes.

In [1]:
## to fix bit length of bit list to n bits
def fix_bit_length(num : str, n : int):
    length = len(num)
    if length < n:
        num = '0'*(n - length) + num
    if len(num) > n:
        num = num[-n:]
    return num

# Bit manipulations
## To invert a single bit
def invert(bit):
    return '1' if bit == '0' else '0'

## To complement all bits
def complement_bits(x:str):
    return ''.join(invert(bit) for bit in x)

# remove msb, add my own bit at lsb
def shift_left(x:str, bit = '0'):
    outbit = x[0]
    x = x[1:] + bit
    return outbit, x

# remove lsb, add my own bit at msb
def shift_right(x:str, bit = '0'):
    outbit = x[-1]
    x = bit+x[:-1]
    return x, outbit

# remove lsb, add msb to msb
def arithmetic_shift_right(x:str):
    outbit = x[-1]
    x = x[0] + x[:-1]
    return x, outbit

# extract msb and lsb
def extract_msb(x:str):
    return x[0]
def extract_lsb(x:str):
    return x[-1]

# bit to logic and vice versa
def bit_to_logic(x:str):
    return x != '0'
def logic_to_bit(x:bool):
    return '1' if x else '0'

# set msb and lsb to specified bit value
def set_msb_to(x:str, bit = '0'):
    x = bit + x[1:]
    return x
def set_lsb_to(x:str, bit = '0'):
    x = x[:-1] + bit
    return x

# Input
## To input two numbers asking bit length:
def input_two_binary_numbers_and_length():
    n = int(input("Please input number of bits = "))
    A = input("Please enter  first binary number = ")
    B = input("Please enter second binary number = ")
    A = fix_bit_length(A, n)
    B = fix_bit_length(B, n)
    return A, B, n

def input_two_binary_numbers_of_n_length(n:int = 4):
    A = input("Please enter  first binary number = ")
    B = input("Please enter second binary number = ")
    A = fix_bit_length(A, n)
    B = fix_bit_length(B, n)
    return A, B

def input_two_binary_numbers_of_length_4():
    return input_two_binary_numbers_of_n_length()

if __name__ == '__main__':
    num = "110101"
    print("Original number  = ", num)
    num = fix_bit_length(num, 10)
    print("Fixing to 10     = ", num)
    num = fix_bit_length(num, 5)
    print("Fixing to 5      = ", num)
    num = complement_bits(num)
    bit = 1
    ibit = invert(bit)
    print(f"Inverse of {bit} is {ibit}")
    iibit = invert(ibit)
    print(f"Inverse of {ibit} is {iibit}")
    print("Inverting bits   = ", num)
    out, num = shift_left(num, '0')
    print("Shifting left    = ", out, num)
    num, out = arithmetic_shift_right(num)
    print("Shifting right arithmatically = ", num, out)
    num, out = shift_right(num)
    print("Shifting right (logically)  = ", num, out)
    print("Finally num      = ", num)
    print(f"MSB of {num} = ", extract_msb(num))
    print(f"LSB of {num} = ", extract_lsb(num))
    num = set_msb_to(num, '1') 
    print("Setting msb to 1 = ", num)
    num = set_lsb_to(num, '0') 
    print("Setting lsb to 0 = ", num)
    
    print()
    for y in [True, False]:
        ybit = logic_to_bit(y)
        print(f"{y} is {ybit}")
    for x in ['1', '0']:
        xlogic = bit_to_logic(x)
        print(f"{x} is {xlogic}")

Original number  =  110101
Fixing to 10     =  0000110101
Fixing to 5      =  10101
Inverse of 1 is 0
Inverse of 0 is 1
Inverting bits   =  01010
Shifting left    =  0 10100
Shifting right arithmatically =  11010 0
Shifting right (logically)  =  01101 0
Finally num      =  01101
MSB of 01101 =  0
LSB of 01101 =  1
Setting msb to 1 =  11101
Setting lsb to 0 =  11100

True is 1
False is 0
1 is True
0 is False


In [2]:
# To check the input functions:
if __name__ == '__main__':
    if 0: # convert to true if I want them to run
        print("Inputting two 4-bit binary numbers")
        A, B = input_two_binary_numbers_of_length_4()
        print(f"A = {A}\nB = {B}")
        print()

        print("Inputting two binary number of length 5")
        n = 5
        A, B = input_two_binary_numbers_of_n_length(n)
        print(f"A = {A}\nB = {B}")
        print()

        print("Inputting two binary numbers by asking length to user:")
        A, B, n = input_two_binary_numbers_and_length()
        print(f"bit_length (n) = {n}\nA = {A}\nB = {B}")
        print()
    

## Addition
1. N bit adder

The following code segment implements the addition's logic.

In [20]:
# to add two bits
def half_adder(a,b):
    x, y = bit_to_logic(a), bit_to_logic(b)
    cy = x and y # carry (ANDing)
    sm = x ^ y # sum (XORing)
    carry = logic_to_bit(cy)
    sum = logic_to_bit(sm)
    return carry, sum

# to add three bits
def full_adder(a,b,c): # using two half adders
    hc1, half_sum = half_adder(a,b)
    hc2, sum = half_adder(half_sum, c)
    carry = '1' if '1' in [hc1, hc2] else '0'
    return carry, sum

    # carry is complicated:
    # half_carry 1 or 2 or 

# same but different implementation
def full_adder_2(a, b, c = '0'): # using normal logic
    a, b, c = [bit_to_logic(x) for x in [a, b, c]]
    carry = (a and b) or (b and c) or (c and a)
    sum = a ^ b ^ c 
    return logic_to_bit(carry), logic_to_bit(sum)

# to add bits of any length
def binary_add(A:str, B:str, n:int, cy='0'):
    A = fix_bit_length(A, n)
    B = fix_bit_length(B, n)
    sum = ''
    for x, y in zip(A[::-1],B[::-1]):
        cy, sm = full_adder(x, y, cy)
        sum = sm + sum
    return cy, sum

if __name__ == '__main__':
    a, b, c = '0', '0', '0'
    cy, sum = full_adder(a, b, c)
    print(f"sum of {(a, b, c)} is {(cy, sum)}")
    # print("Sum of binary_add(a,b,c")")

    A = '0011'
    B = '0111'
    print(f"Sum of {A} and {B} is {binary_add(A, B, 4)}")

sum of ('0', '0', '0') is ('0', '0')
Sum of 0011 and 0111 is ('0', '1010')


In [21]:
# To check the addition
if __name__ == '__main__':
    A, B, n = '1001', '1100', 4
    # uncomment below to manually input vaues
    # A, B, n = input_two_binary_numbers_and_length()

    c, sum = binary_add(A, B, n)
    print(f"  {A}")
    print(f"+ {B}")
    print('-' * (n+4))
    print(f"{c} {sum}")

  1001
+ 1100
--------
1 0101



## Subtraction
2. N bit subtractor

The following code segment implements the n-bit subraction

In [24]:
# To generate 2's complement
def twos_complement(bits:list):
    bits = complement_bits(bits)
    _, bits = binary_add(bits, '1', len(bits))
    return bits

# To perform binary subtraction
def binary_subtract(A:str, B:str, n:int, borrow = '0'):
    if borrow == '0':
        B = twos_complement(B)
    else:
        B = complement_bits(B)
    (b, diff) = binary_add(A, B, n)
    return (invert(b), diff)


if __name__ == '__main__':
    A = '00111'
    B = twos_complement(A)
    print(f"2's complement of {A} is {B}")

    D = binary_subtract(A, B, 5, '1')
    print(f"{A} - (-{A}) -1 = {D}")

2's complement of 00111 is 11001
00111 - (-00111) -1 = ('1', '01101')


In [28]:
# To check the subtraction
if __name__ == '__main__':
    A, B, n = '1110', '1111', 4
    # uncomment below to manually input vaues
    # A, B, n = input_two_binary_numbers_and_length()

    b, difference = binary_subtract(A, B, n)
    print(f"  {A}")
    print(f"- {B}")
    print('-' * (n+4))
    print(f"{b} {difference}")

  1110
- 1111
--------
1 1111


## Multiplication
There are two implementation of multiplications in this lab:
1. To account for +ve numbers only:

Partial product method

2. To account for +ve as well as -ve numbers:

Booth's multiplication method



### Partial product method
3. Partial Product Method of Multiplication

The following segment implements the partial product method of multiplication.

#### Algorithm

    Initially:
    M                 : Multiplicand
    E, AAAA, QQQQ     : (zero), (zeros), (Multiplier)

    loop n times:
        check LSB (of Q):
        if lsb is 1: EA = EA + B
        shift E,A,Q right once, last bit discard
    return A, Q

#### Example

Multiplying 00111 with 01001 in 5 bits.

Initial values:

| Registers | Values | Remarks         |
|-----------|--------|-----------------|
|    M      |  00111 | Multiplicand    |
|    E      |      0 | Initially bit 0 |
|    A      |  00000 | Initially zeros |
|    Q      |  01001 | Multiplier      |

Calculation Table

| Iter | E | A     | Q     | Remarks      |
|-----:|---|-------|-------|--------------|
|    0 | 0 | 00000 | 01001 | Initially    |
|    1 | 0 | 00111 | 01001 | EA -> EA + M |
|      | 0 | 00011 | 10100 | Shift right  |
|    2 | 0 | 00001 | 11010 | Shift right  |
|    3 | 0 | 00000 | 11101 | Shift right  |
|    4 | 0 | 00111 | 11101 | EA -> EA + M |
|      | 0 | 00011 | 11110 | Shift right  |
|    5 | 0 | 00001 | 11111 | Shift right  |

Final answer of multiplication =  0000111111

In [7]:
# Simple multiplication

def binary_partial_product_multiplication(X:str, Y:str, n:int):
    # E,A are zero, M = multiplicand, Q = multipler
    M = fix_bit_length(X, n)
    E = '0'
    A = fix_bit_length('', n)
    Q = fix_bit_length(Y, n)
    
    # Display part (ignore till start of loop)
    print(f"Multiplying {M} with {Q} in {n} bits.")
    print()
    print("Initial values:")
    print()

    # Markdown compatible tabular printing:
    print(f"| Registers | Values | Remarks         |")
    print(f"|-----------|-------{'-'*(n-6)}-|-----------------|")
    print(f"|    M      | {M:>6} | Multiplicand    |")
    print(f"|    E      | {E:>6} | Initially bit 0 |")
    print(f"|    A      | {A:>6} | Initially zeros |")
    print(f"|    Q      | {Q:>6} | Multiplier      |")

    print()
    print("Calculation Table")
    print()
    print(f"| Iter | E | A{' '*n}| Q{' '*n}| Remarks      |")
    print(f"|-----:|---|--{'-'*n}|--{'-'*n}|--------------|")
    print(f"|    0 | {E} | {A} | {Q} | Initially    |")

    # loop n times
    for i in range(n):
        iter = f"{i+1:>4}"
        # check lsb of Q
        if Q[-1] == '1':
            E, A = binary_add(A, M, n)
            print(f"| {iter} | {E} | {A} | {Q} | EA -> EA + M |")
            iter = ' '*4
            
        # shift E,A,Q
        A, bit = shift_right(A, E)
        Q, _ = shift_right(Q, bit)
        print(f"| {iter} | {E} | {A} | {Q} | Shift right  |")
    
    return A+Q

if __name__ == '__main__':
    n = 5
    A, B = '0111', '1001'
    # A,B = input_two_binary_numbers_of_n_length(n)

    mul = binary_partial_product_multiplication(A,B,n)
    print("\nFinal answer of multiplication = ", mul)




Multiplying 00111 with 01001 in 5 bits.

Initial values:

| Registers | Values | Remarks         |
|-----------|--------|-----------------|
|    M      |  00111 | Multiplicand    |
|    E      |      0 | Initially bit 0 |
|    A      |  00000 | Initially zeros |
|    Q      |  01001 | Multiplier      |

Calculation Table

| Iter | E | A     | Q     | Remarks      |
|-----:|---|-------|-------|--------------|
|    0 | 0 | 00000 | 01001 | Initially    |
|    1 | 0 | 00111 | 01001 | EA -> EA + M |
|      | 0 | 00011 | 10100 | Shift right  |
|    2 | 0 | 00001 | 11010 | Shift right  |
|    3 | 0 | 00000 | 11101 | Shift right  |
|    4 | 0 | 00111 | 11101 | EA -> EA + M |
|      | 0 | 00011 | 11110 | Shift right  |
|    5 | 0 | 00001 | 11111 | Shift right  |

Final answer of multiplication =  0000111111


In [8]:
# To check multiplication

### Booth's multiplication method
4. Booth's Multiplication Algorithm

The following segment implements the booth's method of multiplication.

#### Algorithm
    Initially:
    MMMM               : Multiplicand
    AAAA, QQQQ, Q(-1)  : (zeros), (Multiplier), (zero)
*Some implementation use extra bit for A, it's not needed *

    loop n times:
        check LSB of Q, value of Q(-1)
        if:
            10 : A = A - M
            01 : A = A + M
            00/11 : Do nothing
        Arithmatic shift right A,Q,Q(-1)
    return A,Q

#### Example
Multiplying 00111 with 00010 in 5 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   M       |  00111 | Multiplicand     |
|  -M       |  11001 | Neg Multiplicand |
|   A       |  00000 | Initially zeros  |
|   Q       |  00010 | Multiplier       |
|   Q(-1)   |      0 | Initially bit 0  |

Calculation Table

| Iter | A     | Q     | Q(-1) | Remarks       |
|-----:|-------|-------|------:|---------------|
|    0 | 00000 | 00010 |     0 | Initially     |
|    1 | 00000 | 00001 |     0 | ASR [AQQ(-1)] |
|    2 | 11001 | 00001 |     0 | A -> A - M    |
|      | 11100 | 10000 |     1 | ASR [AQQ(-1)] |
|    3 | 00011 | 10000 |     1 | A -> A + M    |
|      | 00001 | 11000 |     0 | ASR [AQQ(-1)] |
|    4 | 00000 | 11100 |     0 | ASR [AQQ(-1)] |
|    5 | 00000 | 01110 |     0 | ASR [AQQ(-1)] |

 Final answer of booth's multiplication =  0000001110


In [9]:
# Booth's multiplication algorithm

def booth_multiplication(X:str, Y:str, n:int):
    # A and Q1 are zero, M = multiplicand, Q = multiplier
    M = fix_bit_length(X, n)
    nM = twos_complement(M)
    A = fix_bit_length('',n)
    Q = fix_bit_length(Y, n)
    Q1 = '0'

    # Display part (ignore till start of the loop)
    print(f"Multiplying {M} with {Q} in {n} bits.")
    print()
    print("Initial values:")

    # Markdown compatible tabular printing:
    print()
    print(f"| Registers | Values | Remarks          |")
    print(f"|-----------|-------{'-'*(n-6)}-|------------------|")
    print(f"|   M       | {M:>6} | Multiplicand     |")
    print(f"|  -M       | {nM:>6} | Neg Multiplicand |")
    print(f"|   A       | {A:>6} | Initially zeros  |")
    print(f"|   Q       | {Q:>6} | Multiplier       |")
    print(f"|   Q(-1)   | {Q1:>6} | Initially bit 0  |")


    print()
    print("Calculation Table")
    print()
    print(f"| Iter | A{' '*n}| Q{' '*n}| Q(-1) | Remarks       |")
    print(f"|-----:|--{'-'*n}|--{'-'*n}|------:|---------------|")
    print(f"|    0 | {A} | {Q} | {Q1:>5} | Initially     |")

    # loop n times
    for i in range(n):
        iter = f"{i+1:>4}"
        # check Q's lsb,Q1
        check_value = Q[-1]+Q1
        if check_value == '01':
            _, A = binary_add(A,M,n)
            print(f"| {iter} | {A} | {Q} | {Q1:>5} | A -> A + M    |")
            iter = ' '*4
        elif check_value == '10':
            _, A = binary_subtract(A,M,n)
            print(f"| {iter} | {A} | {Q} | {Q1:>5} | A -> A - M    |")
            iter = ' '*4
        
        # Arithmetic Shift Right : A,Q,Q(-1)
        A, bit = arithmetic_shift_right(A)
        Q, Q1  = shift_right(Q, bit)
        print(f"| {iter} | {A} | {Q} | {Q1:>5} | ASR [AQQ(-1)] |")
    return A+Q


if __name__ == '__main__':
    
    n = 5
    A,B = '0111', '0010'
    # A,B = input_two_binary_numbers_of_n_length(n)

    mul = booth_multiplication(A, B, n)
    print("\n Final answer of booth's multiplication = ", mul)


Multiplying 00111 with 00010 in 5 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   M       |  00111 | Multiplicand     |
|  -M       |  11001 | Neg Multiplicand |
|   A       |  00000 | Initially zeros  |
|   Q       |  00010 | Multiplier       |
|   Q(-1)   |      0 | Initially bit 0  |

Calculation Table

| Iter | A     | Q     | Q(-1) | Remarks       |
|-----:|-------|-------|------:|---------------|
|    0 | 00000 | 00010 |     0 | Initially     |
|    1 | 00000 | 00001 |     0 | ASR [AQQ(-1)] |
|    2 | 11001 | 00001 |     0 | A -> A - M    |
|      | 11100 | 10000 |     1 | ASR [AQQ(-1)] |
|    3 | 00011 | 10000 |     1 | A -> A + M    |
|      | 00001 | 11000 |     0 | ASR [AQQ(-1)] |
|    4 | 00000 | 11100 |     0 | ASR [AQQ(-1)] |
|    5 | 00000 | 01110 |     0 | ASR [AQQ(-1)] |

 Final answer of booth's multiplication =  0000001110


In [10]:
# To check the multiplication

## Division
There are two implementation for division in this lab:
1. Restoring Division Method
2. Non Restoring Division Method

### Restoring Division
5. Restoring Division

The following segment implements the restoring division

#### Algorithm
    Initially:
    B                  : divisor
    A, Q               : zeros, dividend
    
    loop n times:
        shift left AQ
        A -> A - B
        set the inverse of MSB to LSB
        If MSB is 1 (negative), only then restore A back (A=A+B)
    Remainder is A, Quotient is Q
    Return them

#### Example
Dividing 1011 by  0101 in 4 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   B       |   0101 | Divisor          |
|   A       |   0000 | Initially zeros  |
|   Q       |   1011 | Dividend         |

Calculation Table

| Iter | A    | Q    | Remarks                |
|-----:|------|------|------------------------|
|    0 | 0000 | 1011 | Initially              |
|    1 | 0001 | 0110 | Shift Left AQ          |
|      | 1100 | 0110 | A -> A - B             |
|      | 1100 | 0110 | Setting LSB            |
|      | 0001 | 0110 | Restoring (A -> A + B) |
|    2 | 0010 | 1100 | Shift Left AQ          |
|      | 1101 | 1100 | A -> A - B             |
|      | 1101 | 1100 | Setting LSB            |
|      | 0010 | 1100 | Restoring (A -> A + B) |
|    3 | 0101 | 1000 | Shift Left AQ          |
|      | 0000 | 1000 | A -> A - B             |
|      | 0000 | 1001 | Setting LSB            |
|    4 | 0001 | 0010 | Shift Left AQ          |
|      | 1100 | 0010 | A -> A - B             |
|      | 1100 | 0010 | Setting LSB            |
|      | 0001 | 0010 | Restoring (A -> A + B) |

Remainder = 0001 and Quotient = 0010

In [11]:
# Restoring division algorithm

def restoring_division(X:str, Y:str, n:int):
    # Init values
    B = fix_bit_length(Y, n) # Divisor
    A = fix_bit_length('', n) # Zeros
    Q = fix_bit_length(X, n) # Dividend

    # Display part ignore till loop
    print(f"Dividing {Q} by  {B} in {n} bits.")
    print()
    print("Initial values:")

    # Markdown compatible variable listing in table
    print()
    print(f"| Registers | Values | Remarks          |")
    print(f"|-----------|-------{'-'*(n-6)}-|------------------|")
    print(f"|   B       | {B:>6} | Divisor          |")
    print(f"|   A       | {A:>6} | Initially zeros  |")
    print(f"|   Q       | {Q:>6} | Dividend         |")

    # displaying calculation table
    print()
    print("Calculation Table")
    print()
    print(f"| Iter | A{' '*n}| Q{' '*n}| Remarks                |")
    print(f"|-----:|--{'-'*n}|--{'-'*n}|------------------------|")
    print(f"|    0 | {A} | {Q} | Initially              |")

    # loop n times
    for i in range(n):
        # shift left
        bit, Q = shift_left(Q)
        _, A = shift_left(A, bit)

        # display
        iter = f"{i+1:>4}"
        print(f"| {iter} | {A} | {Q} | Shift Left AQ          |")
        iter = ' '*4

        # subtract
        _, A = binary_subtract(A, B, n)
        print(f"| {iter} | {A} | {Q} | A -> A - B             |")
        msb = A[0]

        # Put the inverse of MSB to LSB
        Q = Q[:-1] + invert(msb)
        print(f"| {iter} | {A} | {Q} | Setting LSB            |")

        if msb == '1':
            _, A = binary_add(A, B, n)
            print(f"| {iter} | {A} | {Q} | Restoring (A -> A + B) |")
    return A, Q


if __name__ == '__main__':
    n = 4
    A, B = '1011', '0101'
    # A, B = input_two_binary_numbers_of_n_length(n)

    remainder, quotient = restoring_division(A,B,n)
    print(f"\nRemainder = {remainder} and Quotient = {quotient}")

Dividing 1011 by  0101 in 4 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   B       |   0101 | Divisor          |
|   A       |   0000 | Initially zeros  |
|   Q       |   1011 | Dividend         |

Calculation Table

| Iter | A    | Q    | Remarks                |
|-----:|------|------|------------------------|
|    0 | 0000 | 1011 | Initially              |
|    1 | 0001 | 0110 | Shift Left AQ          |
|      | 1100 | 0110 | A -> A - B             |
|      | 1100 | 0110 | Setting LSB            |
|      | 0001 | 0110 | Restoring (A -> A + B) |
|    2 | 0010 | 1100 | Shift Left AQ          |
|      | 1101 | 1100 | A -> A - B             |
|      | 1101 | 1100 | Setting LSB            |
|      | 0010 | 1100 | Restoring (A -> A + B) |
|    3 | 0101 | 1000 | Shift Left AQ          |
|      | 0000 | 1000 | A -> A - B             |
|      | 0000 | 1001 | Setting LSB            |
|    4 | 0001 | 0010 | Shift Left AQ          

In [12]:
# To check the restoring division

### Non Restoring Division
6. Non Restoring Division

The following segment implements the non-restoring division

#### Algorithm
    Initially:
    B               : divisor
    A, Q            : zeros, dividend

    loop n times:
        store msb
        shift left AQ
        Add or subtract depending on stored msb equaling to:
            0: A -> A - B
            1: A -> A + B

        extract new msb
        set the inverse of msb to lsb
    
    Remainder is A, Quotient is Q
    if A is negative, restore it : A -> A + B
    Return them
        
#### Example
Dividing 1101 by  0011 in 4 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   B       |   0011 | Divisor          |
|   Bneg    |   0011 | Neg of Divisor   |
|   A       |   0000 | Initially zeros  |
|   Q       |   1101 | Dividend         |

Calculation Table

| Iter | A    | Q    | Remarks                |
|-----:|------|------|------------------------|
|    0 | 0000 | 1101 | Initially              |
|    1 | 0001 | 1010 | Shift Left AQ          |
|      | 1110 | 1010 | A -> A - B             |
|      | 1110 | 1010 | Setting LSB            |
|    2 | 1101 | 0100 | Shift Left AQ          |
|      | 0000 | 0100 | A -> A + B             |
|      | 0000 | 0101 | Setting LSB            |
|    3 | 0000 | 1010 | Shift Left AQ          |
|      | 1101 | 1010 | A -> A - B             |
|      | 1101 | 1010 | Setting LSB            |
|    4 | 1011 | 0100 | Shift Left AQ          |
|      | 1110 | 0100 | A -> A + B             |
|      | 1110 | 0100 | Setting LSB            |
|      | 0001 | 0100 | Restoring (A -> A + B) |

Remainder = 0001 and Quotient = 0100


In [13]:
# Non restoring division algorithm

def non_restoring_division(X:str, Y:str, n:int):
    # Init values
    B = fix_bit_length(Y, n) # Divisor
    Bneg = twos_complement(B) # Negative of divisor
    A = fix_bit_length('', n) # Zeros
    Q = fix_bit_length(X, n) # Dividend

    # Display part ignore till loop
    print(f"Dividing {Q} by  {B} in {n} bits.")
    print()
    print("Initial values:")

    # Markdown compatible variable listing in table
    print()
    print(f"| Registers | Values{''*(n-6)} | Remarks          |")
    print(f"|-----------|-------{'-'*(n-6)}-|------------------|")
    print(f"|   B       | {B:>6} | Divisor          |")
    print(f"|   Bneg    | {B:>6} | Neg of Divisor   |")
    print(f"|   A       | {A:>6} | Initially zeros  |")
    print(f"|   Q       | {Q:>6} | Dividend         |")

    # displaying calculation table
    print()
    print("Calculation Table")
    print()
    print(f"| Iter | A{' '*n}| Q{' '*n}| Remarks                |")
    print(f"|-----:|--{'-'*n}|--{'-'*n}|------------------------|")
    print(f"|    0 | {A} | {Q} | Initially              |")

    # loop n times
    for i in range(n):
        # Store the MSB
        msb = A[0]
        
        # shift left
        bit, Q = shift_left(Q)
        _, A = shift_left(A, bit)

        # display
        iter = f"{i+1:>4}"
        print(f"| {iter} | {A} | {Q} | Shift Left AQ          |")
        iter = ' '*4

        # Subtract or add depending on msb
        if msb == '0': # subtract
            _, A = binary_add(A, Bneg, n)
            print(f"| {iter} | {A} | {Q} | A -> A - B             |")
        else: # add
            _, A = binary_add(A, B, n)
            print(f"| {iter} | {A} | {Q} | A -> A + B             |")

        # Put the inverse of MSB to LSB
        msb = A[0]
        Q = set_lsb_to(Q, invert(msb))
        print(f"| {iter} | {A} | {Q} | Setting LSB            |")

    # check msb and make rem positive by restoring if needed
    msb = A[0]
    if msb == '1':
        _, A = binary_add(A, B, n)
        print(f"| {iter} | {A} | {Q} | Restoring (A -> A + B) |")
    
    return A, Q

if __name__ == '__main__':
    n = 4
    A, B = '1101', '0011'
    # A, B = input_two_binary_numbers_of_n_length(n)

    remainder, quotient = non_restoring_division(A,B,n)
    print(f"\nRemainder = {remainder} and Quotient = {quotient}")

Dividing 1101 by  0011 in 4 bits.

Initial values:

| Registers | Values | Remarks          |
|-----------|--------|------------------|
|   B       |   0011 | Divisor          |
|   Bneg    |   0011 | Neg of Divisor   |
|   A       |   0000 | Initially zeros  |
|   Q       |   1101 | Dividend         |

Calculation Table

| Iter | A    | Q    | Remarks                |
|-----:|------|------|------------------------|
|    0 | 0000 | 1101 | Initially              |
|    1 | 0001 | 1010 | Shift Left AQ          |
|      | 1110 | 1010 | A -> A - B             |
|      | 1110 | 1010 | Setting LSB            |
|    2 | 1101 | 0100 | Shift Left AQ          |
|      | 0000 | 0100 | A -> A + B             |
|      | 0000 | 0101 | Setting LSB            |
|    3 | 0000 | 1010 | Shift Left AQ          |
|      | 1101 | 1010 | A -> A - B             |
|      | 1101 | 1010 | Setting LSB            |
|    4 | 1011 | 0100 | Shift Left AQ          |
|      | 1110 | 0100 | A -> A + B             |
|   

In [14]:
# To check the non restoring division algorithm

## Cache Replacement Algorithm
7. Cache replacement algorithm
There are two cache replacement algorithms we do in this lab: <br />
a. FIFO <br />
b. LRU <br />

### Cache replacement algorithm by FIFO

a. FIFO 

The following segment implements the cache-replacement algorithm by FIFO.

In [15]:
# Cache replacement by FIFO
refString = [1,4,0,3,4,5,6,4, 7 ,9]
pageFrame = 4

def FIFO():
    insertAt = 0
    cache = []
    print("Insert\t","Cache\t")
    for i in refString:
        case = ""
        if i in cache :
            case = "hit"
        else:
            case = "miss"
            if len(cache) < pageFrame :
                cache.append(i)
            else:
                cache[insertAt] = i
            insertAt += 1
            insertAt %= pageFrame
            pass
        print (f"{i}\t{cache}\t{case}")

FIFO()

Insert	 Cache	
1	[1]	miss
4	[1, 4]	miss
0	[1, 4, 0]	miss
3	[1, 4, 0, 3]	miss
4	[1, 4, 0, 3]	hit
5	[5, 4, 0, 3]	miss
6	[5, 6, 0, 3]	miss
4	[5, 6, 4, 3]	miss
7	[5, 6, 4, 7]	miss
9	[9, 6, 4, 7]	miss


### Cache replacement algorithm by LRU
b. LRU

The following segment implements the cache-replacement algorithm by LRU.

In [16]:
# Cache replacement by LRU
refString = [1,4,0,3,4,5,6,4, 7 ,9]
pageFrame = 4

def LRU():
    cache = {}
    print(f"Insert\tCache")
    for i in refString:
        case = ""
        if i in cache:
            case = "hit"
            prev_index = cache[i]
            for j in cache:
                if cache[j] < prev_index:
                    cache[j] += 1
            cache[i] = 0
        else:
            case = "miss"
            for j in cache.copy():
                cache[j] += 1
                if cache[j] == pageFrame:
                    del cache[j]
            cache[i] = 0
            pass
        print (f"{i}\t{cache}\t{case}")

LRU()

Insert	Cache
1	{1: 0}	miss
4	{1: 1, 4: 0}	miss
0	{1: 2, 4: 1, 0: 0}	miss
3	{1: 3, 4: 2, 0: 1, 3: 0}	miss
4	{1: 3, 4: 0, 0: 2, 3: 1}	hit
5	{4: 1, 0: 3, 3: 2, 5: 0}	miss
6	{4: 2, 3: 3, 5: 1, 6: 0}	miss
4	{4: 0, 3: 3, 5: 2, 6: 1}	hit
7	{4: 1, 5: 3, 6: 2, 7: 0}	miss
9	{4: 2, 6: 3, 7: 1, 9: 0}	miss
